core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1538,7 +1538,7 @@ class SparkContext(config: SparkConf) extends Logging {
     val scheme = new URI(schemeCorrectedPath).getScheme
     if (!Array("http", "https", "ftp").contains(scheme)) {
       val fs = hadoopPath.getFileSystem(hadoopConfiguration)
-      val isDir = fs.getFileStatus(hadoopPath).isDirectory
+      val isDir = fs.getFileStatus(new Path(path)).isDirectory
```

Comment:
do you know why this fix the bug?

Reply:
@cloud-fan,  yes ,
org.apache.hadoop.fs.RawLocalFileSystem#deprecatedGetFileStatus (inside this method path.exists will return false)
because java.io.File#exists method will return false (resion:- the file denoted by this abstract pathname is not exists)
becuase space is got encoded by %20 but if will pass simple path then it will find the file and will return true.

Reply:
java.io.FileNotFoundException: File file:/Users/ankitrajboudh/test%20dir/ankit.conf does not exist
  at org.apache.hadoop.fs.RawLocalFileSystem.deprecatedGetFileStatus(RawLocalFileSystem.java:611)
  at org.apache.hadoop.fs.RawLocalFileSystem.getFileLinkStatusInternal(RawLocalFileSystem.java:824)
  at org.apache.hadoop.fs.RawLocalFileSystem.getFileStatus(RawLocalFileSystem.java:601)
  at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:428)
  at org.apache.spark.SparkContext.addFile(SparkContext.scala:1541)
  at org.apache.spark.SparkContext.addFile(SparkContext.scala:1505)

Reply:
in return statement : 
(fs.getBooleanAttributes(this)  = 0 (if path = file:/Users/ankitrajboudh/test%20dir/ankit.conf)
(fs.getBooleanAttributes(this)  != 0 (file:/Users/ankitrajboudh/test dir/ankit.conf)

java.io.File#exists - Please check this exists method return statement- 
public boolean exists() {
        SecurityManager security = System.getSecurityManager();
        if (security != null) {
            security.checkRead(path);
        }
        if (isInvalid()) {
            return false;
        }
        return ((fs.getBooleanAttributes(this) & FileSystem.BA_EXISTS) != 0);
    }

Reply:
Is there any official document mention this? e.g. if we extend the path with scheme, then it can't contain spaces.

Reply:
Yes, I have read the hadoop FileSystem document:
(https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileSystem.html#getFileStatus-org.apache.hadoop.fs.Path-)

In this document they have not mention any limitation 

```
public abstract FileStatus getFileStatus(Path f)
                                  throws IOException
```
As per my knowledge  this `(new File(path).getCanonicalFile.toURI)` will convert **space** to **%20** in path which is causing the problem.

Reply:
it seems the problem is we convert URI to string and then create Path from string. What if we create Path from URI directly?

Reply:
create Path from URI directly is working,  but again we will get exception during fetching schema `val scheme = new URI(schemeCorrectedPath).getScheme` because file path will contain space

```
java.net.URISyntaxException: Illegal character in path at index 25: /Users/ankitrajboudh/test dir/ankit.conf
  at java.net.URI$Parser.fail(URI.java:2848)
  at java.net.URI$Parser.checkChars(URI.java:3021)
  at java.net.URI$Parser.parseHierarchical(URI.java:3105)
  at java.net.URI$Parser.parse(URI.java:3063)
  at java.net.URI.<init>(URI.java:588)
  at org.apache.spark.SparkContext.addFile(SparkContext.scala:1538)
  at org.apache.spark.SparkContext.addFile(SparkContext.scala:1505)
  ... 47 elided
```

 for handling this i we will pass URI.toString `val scheme = new URI(uri.toString).getScheme`